atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/ShardedSweepableBucketRetriever.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * (c) Copyright 2024 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.sweep.asts;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.palantir.atlasdb.sweep.asts.ShardedSweepTimestampManager.SweepTimestamps;
+import com.palantir.atlasdb.sweep.asts.SweepStateCoordinator.SweepableBucket;
+import com.palantir.atlasdb.sweep.queue.ShardAndStrategy;
+import com.palantir.atlasdb.table.description.SweeperStrategy;
+import com.palantir.refreshable.Refreshable;
+import com.palantir.tracing.CloseableTracer;
+import java.time.Duration;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+public final class ShardedSweepableBucketRetriever implements SweepableBucketRetriever {
+    private final Refreshable<Integer> numShards;
+    private final SweeperStrategy strategy;
+
+    private final ShardedRetrievalStrategy shardedRetrievalStrategy;
+    private final ShardedSweepTimestampManager sweepTimestampManager;
+    private final ParallelTaskExecutor parallelTaskExecutor;
+
+    private final Refreshable<Integer> maxParallelism;
+
+    private final Refreshable<Duration> maxBackoff;
+
+    // We could use https://github.com/palantir/safe-thread-local-random instead, but this is to make testing easier.
+    // ThreadLocalRandom because Random does not have a bounded next long implementation
+    private final Supplier<ThreadLocalRandom> random;
```

Comment:
I don't really like this - This is used to "fix" the randomness for testing, but smells like I should clean the design up somewhere.

This is used for the tests `boundedBackoffBetweenRequests` (where we set the random value to be the max value, and assert the sleep is no more than that) and `backoffCanBeInterrupted` (to force a high backoff and show that it can be interrupted)

For `boundedBackoffBetweenRequests`, we could change the assertion from [maxBackoff, maxBackoff + epsilon] to be [1, maxBackoff + epsilon] to match the _actual_ behaviour.

For `backoffCanBeInterrupted`, it's a bit more tricky, in that we could spuriously succeed if the backoff is chosen to be really small, but we could remove the "fixed" randomness.

The tradeoff is that the tests now have a random component, but arguably they might better reflect the guarantees that we actually have.

Reply:
I see. Yeah let's have a supplier of the backoffs / durations?